nomad/task_group_host_volume_claim_endpoint.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package nomad
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	metrics "github.com/hashicorp/go-metrics/compat"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/state/paginator"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// TaskGroupHostVolumeClaim is the server RPC endpoint for task group volume claims
+type TaskGroupHostVolumeClaim struct {
+	srv    *Server
+	ctx    *RPCContext
+	logger hclog.Logger
+}
+
+func NewTaskGroupVolumeClaimEndpoint(srv *Server, ctx *RPCContext) *TaskGroupHostVolumeClaim {
+	return &TaskGroupHostVolumeClaim{srv: srv, ctx: ctx, logger: srv.logger.Named("task_group_host_volume_claim")}
+}
+
+func (tgvc *TaskGroupHostVolumeClaim) List(args *structs.TaskGroupVolumeClaimListRequest, reply *structs.TaskGroupVolumeClaimListResponse) error {
+	authErr := tgvc.srv.Authenticate(tgvc.ctx, args)
+	if done, err := tgvc.srv.forward(structs.TaskGroupHostVolumeClaimListRPCMethod, args, args, reply); done {
+		return err
+	}
+	tgvc.srv.MeasureRPCRate("task_group_volume_claim", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "task_group_volume_claim", "list"}, time.Now())
+
+	allowClaim := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+	aclObj, err := tgvc.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowClaim(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	ns := args.RequestNamespace()
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, stateStore *state.StateStore) error {
+			var iter memdb.ResultIterator
+			var err error
+
+			switch {
+			case args.JobID != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByFields(ws, state.TgvcSearchableFields{Namespace: ns, JobID: args.JobID})
+			case args.VolumeName != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByFields(ws, state.TgvcSearchableFields{Namespace: ns, VolumeName: args.VolumeName})
+			case args.TaskGroup != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByFields(ws, state.TgvcSearchableFields{Namespace: ns, TaskGroupName: args.TaskGroup})
+			default:
+				iter, err = stateStore.GetTaskGroupHostVolumeClaims(ws)
+			}
```

Comment:
These args aren't being used to select an index, and we have a `paginator.Filter` down below. So we have logic being split between the `state` package, this section, and the paginator. I'm pretty sure that's resulting is mishandling of wildcard namespaces here too, as the logic in the `state` package will always filter that out.

I do like the notion of the `TgvcSearchableFields` object though. Rather than checking all these conditionals and then checking most of them again in the `TaskGroupHostVolumeClaimsByFields` and the paginator, we could _always_ populate the `TgvcSearchableFields` struct (including any zero values from the args) and then use that in the `paginator.Filter`. The trouble with that is that you want that logic to be available in the scheduler too, so maybe we end up with something like: `TaskGroupHostVolumeClaimsByFields`.

```suggestion
			iter, err = stateStore.TaskGroupHostVolumeClaimsByFields(ws, 
				state.TgvcSearchableFields{
					Namespace: ns, 
					JobID: args.JobID,
					VolumeName: args.VolumeName,
					TaskGroupName: args.TaskGroup,
				})
```

You'd still have to deal with ACL checking and fix up the wildcard namespace checks.

Reply:
Have a look at [21a47b1](https://github.com/hashicorp/nomad/pull/25114/commits/21a47b114045e582081df5dea3952d2c6798172b). 

I moved wildcard ns checks into `TaskGroupHostVolumeClaimByFields`, removed redundant filters from the iterator construction, and kept the prefix and ACL check in the paginator filter.